library/std/src/thread/local.rs

```diff
@@ -145,12 +145,128 @@ macro_rules! thread_local {
     );
 }
 
+/// Declare a new thread local storage key of type [`std::thread::LocalKey`]
+/// which has a constant initialization expression.
+///
+/// This macro is the same as [`thread_local!`] except that it requires that
+/// the initialization expression is constant rather than allowing it to be
+/// evaluated at runtime. This difference in initialization allows this form of
+/// thread locals to be more efficient on supported platforms and elide a
+/// runtime check as to whether the thread local has already been initialized
+/// when accessed. Otherwise this macro is the same as [`thread_local!`],
+/// notably it also supports types with destructors.
+///
+/// The syntax of this macro is the same as [`thread_local!`] as well.
+///
+/// [`std::thread::LocalKey`]: crate::thread::LocalKey
+/// [`thread_local!`]: crate::thread_local
+#[macro_export]
+#[unstable(feature = "thread_local_const_init", issue = "none")]
+#[allow_internal_unstable(thread_local_internals)]
+macro_rules! thread_local_const_init {
+    () => {};
+    ($(#[$attr:meta])* $vis:vis static $name:ident: $t:ty = $init:expr) => (
+        $crate::__thread_local_inner!($(#[$attr])* $vis $name, $t, const $init);
+    );
+    ($(#[$attr:meta])* $vis:vis static $name:ident: $t:ty = $init:expr; $($rest:tt)*) => (
+        $crate::thread_local_const_init!($(#[$attr])* static $name: $t = $init);
+        $crate::thread_local_const_init!($($rest)*);
+    );
+}
+
 #[doc(hidden)]
 #[unstable(feature = "thread_local_internals", reason = "should not be necessary", issue = "none")]
 #[macro_export]
 #[allow_internal_unstable(thread_local_internals, cfg_target_thread_local, thread_local)]
 #[allow_internal_unsafe]
 macro_rules! __thread_local_inner {
+    // used to generate the `LocalKey` value for `thread_local_const_init!`
+    (@key $t:ty, const $init:expr) => {{
+        unsafe fn __getit() -> $crate::option::Option<&'static $t> {
+            // wasm without atomics maps directly to `static mut`, and dtors
+            // aren't implemented because thread dtors aren't really a thing
+            // on wasm right now
+            #[cfg(all(target_arch = "wasm32", not(target_feature = "atomics")))]
+            {
+                static mut VAL: $t = $init;
+                Some(&VAL)
+            }
+
+            // If the platform has support for `#[thread_local]`, use it.
+            #[cfg(all(
+                target_thread_local,
+                not(all(target_arch = "wasm32", not(target_feature = "atomics"))),
```

Comment:
Does wasm32 declare support for target_thread_local?

Reply:
It does, yes, beacuse it's supported with atomics enabled. Last I checked (a year or two ago) it just didn't behave well in LLVM if atomics was disabled.

Reply:
Probably best to file an issue tracking that.